editor/editor_file_system.cpp

```diff
@@ -1958,6 +1958,11 @@ Error EditorFileSystem::_reimport_file(const String &p_file, const HashMap<Strin
 		Error err = cf->load(p_file + ".import");
 		if (err == OK) {
 			if (cf->has_section("params")) {
+				if (cf->has_section_key("params", "gltf/embedded_image_handling") && !cf->has_section_key("params", "gltf/compatibility_version")) {
+					// If an existing import file is missing the glTF
+					// compatibility version, we need to use version 0.
+					cf->set_value("params", "gltf/compatibility_version", 0);
+				}
```

Comment:
This is quite hacky. GLTF specific code shouldn't spill into EditorFileSystem like this.

Why is it needed? Can't you just assign `compat_version = 0` in `editor_scene_importer_gltf.cpp` if the option is missing?

Reply:
No, because there is no way to distinguish the case of a file without a version (that should use the oldest version) from a missing file (that should use the default version).

We could just always default to the latest version, but that won't automatically fix files from Godot 4.1.

Reply:
Ok, I just had a quick look and indeed EditorFileSystem includes the default parameters in case they're missing:
```cpp
	//mix with default params, in case a parameter is missing

	List<ResourceImporter::ImportOption> opts;
	importer->get_import_options(p_file, &opts);
	for (const ResourceImporter::ImportOption &E : opts) {
		if (!params.has(E.option.name)) { //this one is not present
			params[E.option.name] = E.default_value;
		}
	}
```

IMO that disqualifies this approach if we can't differentiate between new files and old files without a new setting. Adding a hack to EditorFileSystem to workaround this design constraint is not the right approach.

Reply:
EditorFileSystem is the only place we can distinguish these cases, so if we can't set the glTF setting here, we need to set a flag for the glTF code to do it.

However, the entire import pipeline of Godot does not allow this to happen. We can't modify the import options in the glTF importer because the parameter is const.

Reply:
What aaron did is similar to the code we wrote in 3.x (hope this doesn't bring up any stressful memories for Akien) https://github.com/godotengine/godot/commit/d92a1728790bc28405b7f0e047b4c17a9b65b4db
Since this was kind of a last minute change there, it didn't get proper design review, so it shouldn't really justify the approach.... in my mind editor_file_system could be the correct place to override defaults for newly imported files, but ideally an API would be added so each importer can assign its own defaults.

Aha!!! I remember now. The place it should be done is the project setting Import Defaults, since those shouldn't apply retroactively but instead should override only newly imported files, right?

Reply:
I'm not suggesting to add new APIs in this release: I'm suggesting it as future work in 4.3+ to permit cleaning up the importer-specific data being added to editor_file_system.cpp by making an api call instead to ask for defaults (even if this requires going through project settings to change import defaults)

Reply:
I would add a new API to ResourceIimporter like this:
`void ResourceImporter::handle_compatibility(Ref<ConfigFile> & p_config_file)`
which does nothing by default
but in the scene importer you propagate all the way to `EditorSceneFormatImporterGLTF::handle_compatibility`
where you can do all the hacking you want.